### PR TITLE
Add Deephaven ticking table support for matplotlib

### PR DESCRIPTION
**Pull request by mofojed**
- Add ticking support for matplotlib
- Add TableAnimation, TableEventSource helper classes
- TableAnimation is a subclass of matplotlib.animation.Animation: https://matplotlib.org/stable/api/_as_gen/matplotlib.animation.Animation.html#matplotlib.animation.Animation
- TableEventSource matches the interface for event_source: https://github.com/matplotlib/matplotlib/blob/4f5cacfde9024076d9ab0cb6ad1c9a7cf352d5f9/lib/matplotlib/animation.py#L843
  - Added a listener to the Table object and calls the update function with the new data whenever there's a change
  - Annoyingly, they just say `object or optional` there, not saying what actually needs to be added. Inspecting the source code though, there's only calls to `start()`, `stop()`, `add_callback` and `remove_callback`.
- Export an input table with the figure that updates with the revision when the figure has been updated

### CODE REVIEW

File: src/deephaven/plugin/matplotlib/__init__.py [hunk 1]
```diff
@@ -1,10 +1,13 @@
 from deephaven.plugin import Registration
+from deephaven.table_listener import listen
```
Comment (devinrsmith) [inline src/deephaven/plugin/matplotlib/__init__.py:2]: You'll need to declare a dependency on deephaven if you want to do this, but unfortunately we don't publish deephaven yet.

File: src/deephaven/plugin/matplotlib/figure_type.py [hunk 2]
```diff
@@ -1,9 +1,110 @@
...
+    from deephaven import new_table
+    from deephaven.column import string_col, int_col
+    import jpy
+    from deephaven.table_listener import _do_locked
+
+    # We need to get the liveness scope so we can run table operations
+    LivenessScope = jpy.get_type('io.deephaven.engine.liveness.LivenessScope')
+    LivenessScopeStack = jpy.get_type('io.deephaven.engine.liveness.LivenessScopeStack')
+    liveness_scope = LivenessScope(True)
+    LivenessScopeStack.push(liveness_scope)
```
Comment (devinrsmith) [inline src/deephaven/plugin/matplotlib/figure_type.py:58]: I believe there are proper wrapers for this now?

File: src/deephaven/plugin/matplotlib/__init__.py (hunk 1 above)
Comment (mofojed) [reply src/deephaven/plugin/matplotlib/__init__.py:2]: If we don't publish deephaven... then what can I do?

File: src/deephaven/plugin/matplotlib/figure_type.py [hunk 3]
```diff
@@ -1,9 +1,110 @@
...
+        _figure_tables[figure] = _make_input_table(figure)
+    return _figure_tables[figure]
+
+def _export_figure(figure):
+    buf = BytesIO()
+
+    # We need to keep track of the figure while drawing it, or the savefig call triggers our stale callback
+    _exporting_figures.add(figure)
+    figure.savefig(buf, format='PNG', dpi=DPI)
+    _exporting_figures.remove(figure)
```
Comment (devinrsmith) [inline src/deephaven/plugin/matplotlib/figure_type.py:104]: Should probably be in a try / finally.

File: src/deephaven/plugin/matplotlib/figure_type.py [hunk 4]
```diff
@@ -1,9 +1,110 @@
...
+            # Check if we're already drawing this figure, and the stale callback was triggered because of our call to savefig
+            if self in _exporting_figures:
+                return
+            update_revision()
+
+        figure.stale_callback = handle_figure_update
+
+    _do_locked(init_table)
+
+    LivenessScopeStack.pop(liveness_scope)
```
Comment (devinrsmith) [inline src/deephaven/plugin/matplotlib/figure_type.py:89]: Logic should likely be guarded w/ try / finally, although I think the wrappers do this better w/ `with` contextmanagers.

File: src/deephaven/plugin/matplotlib/__init__.py (hunk 1 above)
Comment (devinrsmith) [reply src/deephaven/plugin/matplotlib/__init__.py:2]: At the very least, we need to have a note in the README and update our setup install_requires.

Comment (mofojed) [reply src/deephaven/plugin/matplotlib/__init__.py:2]: I'll update the `install_requires`

File: src/deephaven/plugin/matplotlib/figure_type.py (hunk 2 above)
Comment (jmao-denver) [reply src/deephaven/plugin/matplotlib/figure_type.py:58]: liveness has been wrapped.

File: setup.cfg [hunk 5]
```diff
@@ -25,8 +25,10 @@ package_dir=
     =src
 packages=find_namespace:
 install_requires =
+    deephaven
     deephaven-plugin
     matplotlib
+    jpy
```
Comment (devinrsmith) [inline setup.cfg:31]: This is incorrect - it requires "deephaven-jpy", at least as we have it named now (what you see when you do a pip list on a DH install).

File: src/deephaven/plugin/matplotlib/__init__.py [hunk 6]
```diff
@@ -16,6 +19,141 @@ def init_theme():
...
+        # Extends the _step() method for the Animation class. Used
+        # to get the update information
+        self._last_update = update
+        return super()._step(*args)
+
+    def _draw_frame(self, framedata):
+        data = {}
+        for column in self._columns:
+            data[column] = self._table.j_table.getColumn(column).getDirect()
+        self._func(data, self._last_update, *self._args)
```
Comment (jmao-denver) [inline src/deephaven/plugin/matplotlib/__init__.py:159]: The new Table listener interface provides the updates in a dict with key being the column name and the value being a numpy arrray. Not sure if it makes sense to use it for efficiency purpose

File: src/deephaven/plugin/matplotlib/figure_type.py [hunk 7]
```diff
@@ -13,6 +114,7 @@ def is_type(self, object) -> bool:
         return isinstance(object, Figure)
 
     def to_bytes(self, exporter: Exporter, figure: Figure) -> bytes:
-        buf = BytesIO()
-        figure.savefig(buf, format='PNG', dpi=144)
-        return buf.getvalue()
+        input_table, liveness_scope = _get_input_table(figure)
+        exporter.reference(input_table)
+        exporter.reference(liveness_scope)
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply src/deephaven/plugin/matplotlib/figure_type.py:119]: I don't see when 'liveness_scope' gets closed. But it seems to me that passing liveness scope around isn't the typical use pattern.

**Review by jmao-denver — CHANGES_REQUESTED** [review]

File: src/deephaven/plugin/matplotlib/__init__.py (hunk 6 above)
Comment (mofojed) [reply src/deephaven/plugin/matplotlib/__init__.py:159]: Yea, it depends on how the data is used in matplotlib. Different charts use it different. As such I decided to provide both (direct data and the update).

File: src/deephaven/plugin/matplotlib/figure_type.py (hunk 2 above)
Comment (mofojed) [reply src/deephaven/plugin/matplotlib/figure_type.py:58]: Where do I import liveness_scope from? What objects do I need to keep around? What do I do when I'm done with it? What exactly is a liveness scope?

I'm a little confused by it's usage/purpose. Any docs or diagram would be very helpful.

Comment (jmao-denver) [reply src/deephaven/plugin/matplotlib/figure_type.py:57]: Hi, Mike, the liveness scope wrapping was merged early last week. I don't think DevRel has gotten around to document the usage yet. So all we have now is the generated Python doc.
       https://deephaven.io/core/pydoc/code/deephaven.liveness_scope.html.

File: setup.cfg [hunk 8]
```diff
@@ -25,6 +25,8 @@ package_dir=
     =src
 packages=find_namespace:
 install_requires =
+    deephaven
+    deephaven-jpy
```
Comment (mofojed) [inline setup.cfg:29]: @devinrsmith if I add this I get errors when installing the wheel:
```Processing /tmp/deephaven_plugin_matplotlib-0.0.1.dev7-py3-none-any.whl
Collecting deephaven-plugin
  Using cached deephaven_plugin-0.2.0-py3-none-any.whl (7.7 kB)
ERROR: Could not find a version that satisfies the requirement deephaven-jpy (from deephaven-plugin-matplotlib) (from versions: none)
ERROR: No matching distribution found for deephaven-jpy
```

What's the correct course of action here?

Comment (devinrsmith) [reply setup.cfg:29]: How / where are you trying to install it? It should only be installed in a location that already has deephaven setup, yeah?

File: src/deephaven/plugin/matplotlib/figure_type.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mofojed) [reply src/deephaven/plugin/matplotlib/figure_type.py:120]: As discussed on yesterday's call, I probably do not even need liveness scope.

File: src/deephaven/plugin/matplotlib/figure_type.py (hunk 4 above)
Comment (mofojed) [reply src/deephaven/plugin/matplotlib/figure_type.py]: No more Liveness needed

**Review by jmao-denver — APPROVED** [review]
LGTM

**Review by devinrsmith — APPROVED** [review]

**Comment by mofojed** [comment]
@devinrsmith I need you to merge, I do not have permission.

**Pull request merged**